client/src/components/Donation/stripe-card-form.tsx

```diff
@@ -0,0 +1,159 @@
+import {
+  CardNumberElement,
+  CardExpiryElement,
+  useStripe,
+  useElements,
+  Elements
+} from '@stripe/react-stripe-js';
+import { loadStripe } from '@stripe/stripe-js';
+import type {
+  Token,
+  StripeCardNumberElementChangeEvent,
+  StripeCardExpiryElementChangeEvent
+} from '@stripe/stripe-js';
+import React, { useState } from 'react';
+
+import envData from '../../../../config/env.json';
+import { AddDonationData } from './PaypalButton';
+
+const { stripePublicKey }: { stripePublicKey: string | null } = envData;
+interface FormPropTypes {
+  onDonationStateChange: (donationState: AddDonationData) => void;
+  postStripeCardDonation: (token: Token) => void;
+  t: (label: string) => string;
+  theme: string;
+}
+
+interface Element {
+  elementType: 'cardNumber' | 'cardExpiry';
+  complete: boolean;
+  error?: null | { type: 'validation_error'; code: string; message: string };
+}
+
+type PaymentInfoValidation = Element[];
+
+const StripeCardForm = ({
+  theme,
+  t,
+  onDonationStateChange,
+  postStripeCardDonation
+}: FormPropTypes): JSX.Element => {
+  const [isSubmissionValid, setSubmissionValidity] = useState(true);
+  const [isSubmitting, setSubmitting] = useState(false);
+  const [paymentInfoValidation, setPaymentValidity] =
+    useState<PaymentInfoValidation>([
+      {
+        elementType: 'cardNumber',
+        complete: false,
+        error: null
+      },
+      {
+        elementType: 'cardExpiry',
+        complete: false,
+        error: null
+      }
+    ]);
+  const isPaymentInfoValid = paymentInfoValidation.every(
+    ({ complete, error }) => complete && !error
+  );
+
+  const stripe = useStripe();
+  const elements = useElements();
+
+  function handleInputChange(
+    event:
+      | StripeCardNumberElementChangeEvent
+      | StripeCardExpiryElementChangeEvent
+  ) {
+    const { elementType, error, complete } = event;
+    setPaymentValidity(
+      paymentInfoValidation.map(element => {
+        if (element.elementType === elementType)
+          return { elementType, error, complete };
+        return element;
+      })
+    );
+  }
+
+  const options = {
+    style: {
+      base: {
+        fontSize: '18px',
+        color: `${theme === 'night' ? '#fff' : '#0a0a23'}`
+      }
+    }
+  };
+
+  const handleSubmit = async (e: React.FormEvent) => {
+    e.preventDefault();
+
+    if (!isPaymentInfoValid) return setSubmissionValidity(false);
+    else setSubmissionValidity(true);
+
+    setSubmitting(true);
+    if (!isSubmitting && stripe && elements) {
+      const cardElement = elements.getElement(CardNumberElement);
+      if (cardElement) {
+        const { error, token } = await stripe.createToken(cardElement);
+        if (error) {
+          onDonationStateChange({
+            redirecting: false,
+            processing: false,
+            success: false,
+            error: t('donate.went-wrong')
+          });
+        } else if (token) postStripeCardDonation(token);
```

Comment:
Do we need to `await` this or can we just move on to `setSubmitting(false)`?

Reply:
We should, I am yet to find an elegant way to await an action.
On the other hand the saga takes the leading action only in the case of multiple submissions so we should be good.

Reply:
If I understand this correctly, the point is to prevent multiple submissions - and, yeah, `takeLeading` should do that for us.  

All this does is disable the block and prevent additional attempts while it's creating a token.  So, I'd recommend either renaming it to isCreatingStripeToken (because that's what it seems to track) or use the redux store to track the submission state.  That would be set back to false via `postChargeStripeCardComplete`'s reducer.

Does that make sense?

Reply:
Yes, now it checks if the payment is being processed and if the token is being made.